inference-engine/src/transformations/src/transformations/op_conversions/einsum_decomposition.cpp

```diff
@@ -0,0 +1,636 @@
+/// iff. the corresponding label is met in neither the output subscript nor the input subscripts
+/// excluding ones specified by a vector excluded_indices
+///
+/// \param      input_subscripts         The vector of the input subscripts
+/// \param      output_subscript         The output subscript
+/// \param      label_to_check           A label that corresponds to dimension to check
+/// \param      excluded_indices         A vector of input subscript indices to be excluded
+///
+/// \return     true - a dimension to reduce, false - otherwise
+///
+bool is_dimension_reduced(const std::vector<std::string>& input_subscripts, const std::string& output_subscript, const std::string label_to_check,
+                          const std::vector<size_t>& excluded_indices) {
+    for (size_t input_ind = 0; input_ind < input_subscripts.size(); ++input_ind) {
+        const auto& input_subscript = input_subscripts[input_ind];
+        // the subscript is checked only if its index is not in excluded indices list
+        bool check_subscript = (std::find(excluded_indices.begin(), excluded_indices.end(), input_ind) == excluded_indices.end());
+        if (check_subscript && input_subscript.find(label_to_check) != std::string::npos) {
+            return false;
+        }
+    }
+    if (output_subscript.find(label_to_check) != std::string::npos) {
+        return false;
+    }
+    return true;
+}
+
+/// \brief      Generate an input subscript that provides to group dimensions into the common,
+/// separate and reduced dimensions after transpose
+///
+/// \param      input_subscripts         A vector of the input subscripts
+/// \param      common_labels_inds       A vector of indices of the common dimensions
+/// \param      separate_labels_inds     A vector of indices of the separate dimensions
+/// \param      reduced_labels_inds      A vector of indices of the reduced dimensions
+/// \param      is_separate_first        A boolean flag. It is true if the separate dimensions
+/// goes before the reduced dimensions
+///
+/// \return     An input subscript for grouping dimensions
+///
+std::string generate_grouping_subscript(const std::string& input_subscript, const std::vector<int64_t>& common_labels_inds,
+                                        const std::vector<int64_t>& separate_labels_inds, const std::vector<int64_t>& reduced_labels_inds,
+                                        bool& is_separate_first) {
+    // transpose is not needed if common labels, reduced labels and separate labels indices go
+    // concurrently
+    bool is_transpose_needed = false;
+    std::vector<int64_t> labels_inds = common_labels_inds;
+    labels_inds.insert(labels_inds.end(), reduced_labels_inds.begin(), reduced_labels_inds.end());
+    labels_inds.insert(labels_inds.end(), separate_labels_inds.begin(), separate_labels_inds.end());
+    int64_t check_index = 0;
+    for (auto index : labels_inds) {
```

Comment:
Yes, it would be more convenient to `std::generate` for the range generation (like here https://en.cppreference.com/w/cpp/algorithm/generate) and to compare with it reference then . But I am afraid that centos platform in CI that still contains g++ 4.8 compiler will fail. So I have to use this primitive way.